input-methods: add documentation

ericsagnes opened the pull request:
Add basic documentation for input methods.
Address #13265.

ericsagnes committed:
manual: add chapter on input methods

joachifm commented:
I like this. It's to the point and practical, makes sense even to a non-user. If I were to pick a nit (obligatory, right?), it'd be that `nix-env -f '<nixpkgs>'` is strictly unnecessary (or ought to be).

ericsagnes commented:
@joachifm Thanks for the review, I changed the `nix-env` part for a more descriptive text introducing each engine and updated the commit subject.
It might be worth to add something about manual related commits in the [manual](https://nixos.org/nixpkgs/manual/#chap-submitting-changes) to have subject consistency for future commits.

edolstra commented on nixos/doc/manual/default.nix:
```diff
@@ -58,6 +58,7 @@ let
       cp ${../../modules/services/databases/postgresql.xml} configuration/postgresql.xml
       cp ${../../modules/services/misc/gitlab.xml} configuration/gitlab.xml
       cp ${../../modules/security/acme.xml} configuration/acme.xml
+      cp ${../../modules/i18n/inputMethod/default.xml} configuration/input-methods.xml
```
No camelCase in filenames please.

ericsagnes replied:
Thanks, done in the `input-method module: fix folder case` commit.

edolstra commented on nixos/modules/i18n/inputMethod/default.xml:
```diff
@@ -0,0 +1,99 @@
+  <listitem><para>Fcitx: A customizable lightweight input method.</para></listitem>
+  <listitem><para>Nabi: A Korean input method based on XIM.</para></listitem>
+  <listitem><para>Uim: The universal input method, is a library with a XIM bridge.</para></listitem>
+</itemizedlist>
```
I don't really know what this means (in fact I've never heard of these input methods). What about (say) `evdev`

ericsagnes replied:
I have very basic knowledge of `evdev`, but as far as I can tell input methods do different things than `evdev`.
Input methods are mainly used with Chinese, Japanese and Korean languages.
To take example of japanese, that has thousands of characters and no spaces, the input method has to do natural language processing to suggest a list of candidates that matches the inputted text and let the user chose the right one. So for example if you input `umanomimininennbutu` and hit space, the input method will automagically convert that to `馬の耳に念仏` (hitting space more times will show different candidates).

edolstra commented on nixos/modules/i18n/inputMethod/default.xml:
```diff
@@ -0,0 +1,99 @@
+
+<programlisting>
+i18n.inputMethod = {
+  enabled = "fcitx";
```
Why is this `enabled` option not a boolean?

ericsagnes replied:
The motivation was that only one method can be enabled at a time and it was feeling cleaner to have one option to chose which input method to enable, than one boolean `enable` for each with checks that others input methods are not enabled.

The `enabled` option is defined as:

```
  options = {
    i18n.inputMethod = {
      enabled = mkOption {
        type    = types.nullOr (types.enum [ "ibus" "fcitx" "nabi" "uim" ]);
        default = null;
        example = "fcitx";
        description = ''
          Select the enabled input method. Input methods is a software to input symbols that are not available on standard input devices.

          Input methods are specially used to input Chinese, Japanese and Korean characters.

          Currently the following input methods are available in NixOS:

          <itemizedlist>
          <listitem><para>ibus: The intelligent input bus, extra input engines can be added using <literal>i18n.inputMethod.ibus.engines</literal>.</para></listitem>
          <listitem><para>fcitx: A customizable lightweight input method, extra input engines can be added using <literal>i18n.inputMethod.fcitx.engines</literal>.</para></listitem>
          <listitem><para>nabi: A Korean input method based on XIM. Nabi doesn't support Qt 5.</para></listitem>
          <listitem><para>uim: The universal input method, is a library with a XIM bridge. uim mainly support Chinese, Japanese and Korean.</para></listitem>
          </itemizedlist>
        '';
      };
    };
  };
```

It also happen to be a useful central point to manage input methods related files like gtk cache files in #14568.

edolstra commented on nixos/modules/i18n/inputMethod/default.xml:
```diff
@@ -0,0 +1,99 @@
+
+<title>Input Methods</title>
+
+<para>Input methods are an operating system component that allows any data, such as keyboard strokes or mouse movements, to be received as input. In this way users can enter characters and symbols not found on their input devices. Using an input method is obligatory for any language that has more graphemes than there are keys on the keyboard.</para>
```
Please wrap at 80 characters or so.

ericsagnes replied:
Thanks, I updated the commit.

ericsagnes committed:
input-method module: fix folder case

ericsagnes commented:
Thank you!

joachifm commented on nixos/modules/rename.nix:
```diff
@@ -63,7 +63,7 @@ with lib;
 
     # ibus
-    (mkRenamedOptionModule [ "programs" "ibus" "plugins" ] [ "i18n" "inputMethod" "ibus" "engines" ])
+    (mkRenamedOptionModule [ "programs" "ibus" "plugins" ] [ "i18n" "input-method" "ibus" "engines" ])
```
Ah, I didn't notice this one; this won't work, because you didn't actually change the option name (it is independent of the folder name).

ericsagnes replied:
@joachifm Should I make a new PR to fix that or can it be fixed by amending the commit?

joachifm replied:
I pushed a hotfix, no worries

joachifm replied:
This really is the kind of error that travis should pick up on ... oh well

ericsagnes replied:
Thank you!
